### PR TITLE
feat(selectors): allow running selectors in the main world

### LINKED ISSUE #1316: [Feature] Allow custom selector engines to run in the main world

### PR DESCRIPTION
**Pull request by dgozman**
Fixes #1316.

### CODE REVIEW

**Review by pavelfeldman — CHANGES_REQUESTED** [review]
Could you split it into preliminary moves and the actual changes? Otherwise, I don't think I can follow the changes!

File: docs/api.md [hunk 1]
```diff
@@ -3301,14 +3301,16 @@ Contains the URL of the response.
...
 <!-- GEN:stop -->
 
-#### selectors.register(name, script)
+#### selectors.register(name, script[, options])
 - `name` <[string]> Name that is used in selectors as a prefix, e.g. `{name: 'foo'}` enables `foo=myselectorbody` selectors. May only contain `[a-zA-Z0-9_]` characters.
 - `script` <[function]|[string]|[Object]> Script that evaluates to a selector engine instance.
   - `path` <[string]> Path to the JavaScript file. If `path` is a relative path, then it is resolved relative to [current working directory](https://nodejs.org/api/process.html#process_process_cwd).
   - `content` <[string]> Raw script content.
+- `options` <[Object]>
+  - `isolated` <[boolean]> Whether to run this selector engine in isolated JavaScript environment. This environment has access to the same DOM, but not any JavaScript objects from the frame's scripts. Defaults to `true`.
```
[the remark this answers is not in the document]
Comment (dgozman) [reply docs/api.md]: <placeholder asking for a better alternative>

**Comment by dgozman** [comment]
> Could you split it into preliminary moves and the actual changes? Otherwise, I don't think I can follow the changes!

Sure, please see #1536.

**Comment by dgozman** [comment]
Please take another look.

File: docs/api.md [hunk 2]
```diff
@@ -3301,14 +3301,16 @@ Contains the URL of the response.
...
 <!-- GEN:stop -->
 
-#### selectors.register(name, script)
+#### selectors.register(name, script[, options])
 - `name` <[string]> Name that is used in selectors as a prefix, e.g. `{name: 'foo'}` enables `foo=myselectorbody` selectors. May only contain `[a-zA-Z0-9_]` characters.
 - `script` <[function]|[string]|[Object]> Script that evaluates to a selector engine instance.
   - `path` <[string]> Path to the JavaScript file. If `path` is a relative path, then it is resolved relative to [current working directory](https://nodejs.org/api/process.html#process_process_cwd).
   - `content` <[string]> Raw script content.
+- `options` <[Object]>
+  - `isolated` <[boolean]> Whether to run this selector engine in isolated JavaScript environment. This environment has access to the same DOM, but not any JavaScript objects from the frame's scripts. Defaults to `false`. Note that isolation is not guaranteed when this engine is used together with other registered engines.
```
Comment (pavelfeldman) [inline docs/api.md]: s/isolated/contentScript/ ?

**Review by pavelfeldman — APPROVED** [review]

File: docs/api.md (hunk 2 above)
Comment (dgozman) [reply docs/api.md]: Done.

**Pull request merged**